code/go/0chain.net/smartcontract/dbs/event/provider.go

```diff
@@ -108,6 +122,311 @@ func (edb *EventDb) ReplicateProviderAggregates(round int64, limit int, offset i
 	return nil
 }
 
+func (edb *EventDb) BuildChangedProvidersMapFromEvents(events []Event) (ProvidersMap, error) {
+	ids, err := extractIdsFromEvents(events)
+	if err != nil {
+		return nil, err
+	}
+
+	providers := ProvidersMap{
+		spenum.Blobber:    make(map[string]IProvider),
+		spenum.Miner:      make(map[string]IProvider),
+		spenum.Sharder:    make(map[string]IProvider),
+		spenum.Authorizer: make(map[string]IProvider),
+		spenum.Validator:  make(map[string]IProvider),
+	}
+
+	idsLists := map[spenum.Provider][]string{
+		spenum.Blobber:    make([]string, 0, len(ids[spenum.Blobber])),
+		spenum.Miner:      make([]string, 0, len(ids[spenum.Miner])),
+		spenum.Sharder:    make([]string, 0, len(ids[spenum.Sharder])),
+		spenum.Authorizer: make([]string, 0, len(ids[spenum.Authorizer])),
+		spenum.Validator:  make([]string, 0, len(ids[spenum.Validator])),
+	}
+	
+	for provider, ids := range ids {
+		for id := range ids {
+			idsLists[provider] = append(idsLists[provider], id)
+		}
+	}
+
+	providersLists := make(map[spenum.Provider][]IProvider)
+	for provider, ids := range idsLists {
+		providersLists[provider], err = edb.GetProvidersByIds(provider, ids)
+		if err != nil {
+			return nil, err
+		}
+	}
+	for ptype, providersList := range providersLists {
+		for _, provider := range providersList {
+			providers[ptype][provider.GetID()] = provider
+		}
+	}
+
+	return providers, nil
+}
+
+func (edb *EventDb) GetProvidersByIds(ptype spenum.Provider, ids []string) ([]IProvider, error) {
+	switch ptype {
+	case spenum.Blobber:
+		return getProvidersById[*Blobber](edb, ids)
+	case spenum.Miner:
+		return getProvidersById[*Miner](edb, ids)
+	case spenum.Sharder:
+		return getProvidersById[*Sharder](edb, ids)
+	case spenum.Authorizer:
+		return getProvidersById[*Authorizer](edb, ids)
+	case spenum.Validator:
+		return getProvidersById[*Validator](edb, ids)
+	}
+
+	return nil, common.NewError("invalid_provider_type", "invalid provider type")
+}
+
+func getProvidersById[P IProvider](edb *EventDb, ids []string) ([]IProvider, error) {
+	var (
+		model P
+		providers []P
+		tableName = model.TableName()
+	)
+	err := edb.Get().
```

Comment:
how many providers can we have here? should we batch the results to save the memory?

Reply:
Not much, depends on the number of events in one round.